src/DI/ServiceLookup/CallSiteFactory.cs

```diff
@@ -118,12 +114,13 @@ private IServiceCallSite TryCreateOpenGeneric(Type serviceType, ISet<Type> callS
             return null;
         }
 
-        private IServiceCallSite TryCreateEnumerable(Type serviceType, ISet<Type> callSiteChain)
+        private IServiceCallSite TryCreateEnumerable(Type serviceType, CallSiteChain callSiteChain)
         {
             if (serviceType.IsConstructedGenericType &&
                 serviceType.GetGenericTypeDefinition() == typeof(IEnumerable<>))
             {
                 var itemType = serviceType.GenericTypeArguments.Single();
+                callSiteChain.Add(serviceType);
```

Comment:
Hmm, this was not in the chain before, but now it is, so more types are in the circular dependency check. Is this correct? I'm worried about a potential breaking change here because now more checks will be happening. (But, maybe the code was always wrong, and now just a better exception will be thrown.)

Reply:
If I understand everything correctly with this changes we will have slightly less checks, because currently type added to chain before we decide how to resolve it (Use Instance / Call Constructor / Create Enumerable / Run Factory). Instance Use and Factory Run can't cause circular reference so I propose add type to chain only in cases of Enumerable Creation and Constructor Call (with implementation type for more detailed error).

Reply:
`IEnumerable<>` it was, it was just added here: https://github.com/aspnet/DependencyInjection/pull/592/files/bac0c91230d57467eb46c2d1ff8da4ee35f7018a#diff-a7622bce2aa4a5b7dbb9e510b8fe8f86L80

Now instead of adding in one common point it gets added in multiple. Mostly because more information is required (implementation type)

Reply:
Got it, thanks for the explanations.